src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProviderSource.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks.Dataflow;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Order
+{
+    /// <summary>
+    /// Setup the dataflow for a provider that updates solution tree item properties with 
+    /// display order metadata derived from IOrderedSourceItemsDataSourceService
+    /// </summary>
+    [Export(typeof(IProjectTreePropertiesProviderDataSource))]
+    [AppliesTo(ProjectCapability.SortByDisplayOrder)]
+    internal class TreeItemOrderPropertyProviderSource : ChainedProjectValueDataSourceBase<IProjectTreePropertiesProvider>, IProjectTreePropertiesProviderDataSource
+    {
+        private UnconfiguredProject _project;
+        private TreeItemOrderPropertyProvider latestTreeItemOrderPropertyProvider;
+
+        [ImportingConstructor]
+        public TreeItemOrderPropertyProviderSource(UnconfiguredProject project)
+            : base(project.Services)
+        {
+            _project = project;
+        }
+
+        [Import(ExportContractNames.Scopes.UnconfiguredProject)]
+        private IOrderedSourceItemsDataSourceService OrderedItemSource { get; set; }
+
+        protected override IDisposable LinkExternalInput(ITargetBlock<IProjectVersionedValue<IProjectTreePropertiesProvider>> targetBlock)
+        {
+            JoinUpstreamDataSources(OrderedItemSource);
+            
+            var providerProducerBlock = new TransformBlock<IProjectVersionedValue<IReadOnlyCollection<ProjectItemIdentity>>, IProjectVersionedValue<IProjectTreePropertiesProvider>>(
+                orderedItems =>
+                {
+                    if (latestTreeItemOrderPropertyProvider?.OrderedItems != orderedItems.Value)
```

Comment:
Doesn't this get called back for changes that only contain version changes? If so, we're doing a bunch of work for basically nothing.

Reply:
I observe that this callback is called multiple times (e.g. if you add a package reference or make other project changes). However this check is always false unless you change the file display order, so we avoid the expensive code path. It seems this is how this API is intended to work.

---
In reply to: [149219607](https://github.com/dotnet/project-system/pull/2930#discussion_r149219607) [](ancestors = 149219607)